cmake/HPX_GeneratePackageUtils.cmake

```diff
@@ -0,0 +1,250 @@
+# Copyright (c) 2014 Thomas Heller
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+# Replace the "-NOTFOUND" by empty var in case the property is not found
+macro(get_target_property var target property)
+  _get_target_property(${var} ${target} ${property})
+  list(FILTER ${var} EXCLUDE REGEX "-NOTFOUND$")
+endmacro(get_target_property)
+
+# https://github.com/boost-cmake/bcm/blob/master/share/bcm/cmake/BCMPkgConfig.cmake
+# https://gitlab.kitware.com/cmake/cmake/issues/17984
+
+# Recursively add the interface_include_dirs of the dependencies and link them
+function(hpx_collect_usage_requirements target compile_definitions compile_options
+    pic_option include_directories system_include_directories link_libraries
+    link_options already_processed_targets is_component)
+
+  # Check if the target has already been processed
+  list(FIND ${already_processed_targets} ${target} _found)
+  if (NOT (${_found} EQUAL -1))
+    return()
+  endif()
+
+  set(_already_processed_targets ${${already_processed_targets}} ${target})
+
+  get_target_property(_target_compile_definitions ${target} INTERFACE_COMPILE_DEFINITIONS)
+  get_target_property(_target_compile_options ${target} INTERFACE_COMPILE_OPTIONS)
+  get_target_property(_target_pic_option ${target} INTERFACE_POSITION_INDEPENDENT_CODE)
+  get_target_property(_target_include_directories ${target} INTERFACE_INCLUDE_DIRECTORIES)
+  get_target_property(_target_system_include_directories ${target} INTERFACE_SYSTEM_INCLUDE_DIRECTORIES)
+  get_target_property(_target_link_libraries ${target} INTERFACE_LINK_LIBRARIES)
+  get_target_property(_target_link_options ${target} INTERFACE_LINK_OPTIONS)
+  get_target_property(_target_type ${target} TYPE)
+
+  # If the target is a library link against it.
+  if(${_target_type} STREQUAL "STATIC_LIBRARY" OR ${_target_type} STREQUAL "SHARED_LIBRARY")
+    # Necessary cause TARGET_FILE doesn't find the target in the install dir
+    if (${is_component})
+      # We put the link directory and let the user specify -l<component>
+      set(_libraries
+        -L$<INSTALL_INTERFACE:${CMAKE_INSTALL_LIBDIR}>$<BUILD_INTERFACE:$<TARGET_FILE_DIR:${target}>>)
+    else()
+      set(_libraries
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_LIBDIR}/$<TARGET_FILE_NAME:${target}>>$<BUILD_INTERFACE:$<TARGET_FILE:${target}>>)
+    endif()
+  else()
+    set(_libraries "")
+  endif()
+
+  # In case of components no need to do the recursive search
+  if (NOT ${is_component})
+    foreach(dep IN LISTS _target_link_libraries)
+
+      if(${dep} MATCHES "^\\$<LINK_ONLY:([^>]+)>$")
+        # This a private link dependency. Do not inherit the target's usage requirements.
+
+        set(dep_target ${CMAKE_MATCH_1})
+        get_target_property(dep_type ${dep_target} TYPE)
+        if(${dep_type} STREQUAL "STATIC_LIBRARY" OR ${dep_type} STREQUAL "SHARED_LIBRARY")
```

Comment:
I think this needs to be able to handle an empty `dep_type`. Quotes around `${dep_type}` might be enough, otherwise it needs a separate branch. See here: https://cdash.cscs.ch//viewConfigure.php?buildid=70413.